zkevm-circuits/src/circuit_tools.rs

```diff
@@ -0,0 +1,1015 @@
+            None
+        } else {
+            Some(and::expr(self.conditions.iter()))
+        }
+    }
+
+    pub(crate) fn get_condition_expr(&self) -> Expression<F> {
+        self.get_condition().unwrap_or_else(|| 1.expr())
+    }
+
+    pub(crate) fn lookup_table(
+        &mut self,
+        description: &'static str,
+        tag: String,
+        values: Vec<Expression<F>>,
+    ) {
+        let condition = self.get_condition_expr();
+        self.lookup_tables.push(LookupData {
+            description,
+            tag,
+            condition,
+            values,
+        });
+    }
+
+    pub(crate) fn lookup(
+        &mut self,
+        description: &'static str,
+        tag: String,
+        values: Vec<Expression<F>>,
+    ) {
+        let condition = self.get_condition_expr();
+        self.lookups.push(LookupData {
+            description,
+            tag,
+            condition,
+            values,
+        });
+    }
+
+    pub(crate) fn get_lookups<S: AsRef<str>>(&self, tags: &[S]) -> Vec<LookupData<F>> {
+        self.lookups
+            .iter()
+            .cloned()
+            .filter(|lookup| tags.iter().any(|tag| lookup.tag == tag.as_ref()))
+            .collect::<Vec<_>>()
+    }
+
+    pub(crate) fn get_lookup_table<S: AsRef<str>>(&self, tag: S) -> Vec<Expression<F>> {
+        let lookups = self
+            .lookup_tables
+            .iter()
+            .filter(|lookup| lookup.tag == tag.as_ref())
+            .collect::<Vec<_>>();
+        let selector = sum::expr(lookups.iter().map(|lookup| lookup.condition.expr()));
+
+        let mut table = vec![0.expr(); lookups[0].values.len()];
+        for (idx, value) in table.iter_mut().enumerate() {
+            *value = sum::expr(
+                lookups
+                    .iter()
+                    .map(|lookup| selector.expr() * lookup.values[idx].expr()),
+            );
+        }
+        table
+    }
+
+    pub(crate) fn print_stats(&self) {
+        let mut expressions = self.constraints.clone();
+        expressions.sort_by(|a, b| a.1.degree().cmp(&b.1.degree()));
+        for (name, expr) in expressions.iter() {
+            println!("'{}': {}", name, expr.degree());
+        }
+    }
+}
+
+pub(crate) fn merge_lookups<F: FieldExt>(
+    cb: &mut ConstraintBuilder<F>,
+    lookups: Vec<LookupData<F>>,
+) -> (Expression<F>, Vec<Expression<F>>) {
+    merge_values(
+        cb,
+        lookups
+            .iter()
+            .map(|lookup| (lookup.condition.clone(), lookup.values.clone()))
+            .collect::<Vec<_>>(),
+    )
+}
+
+pub(crate) fn merge_values<F: FieldExt>(
+    cb: &mut ConstraintBuilder<F>,
+    values: Vec<(Expression<F>, Vec<Expression<F>>)>,
+) -> (Expression<F>, Vec<Expression<F>>) {
+    let selector = sum::expr(values.iter().map(|(condition, _)| condition.expr()));
+    // Sanity checks (can be removed, here for safety)
+    crate::circuit!([meta, cb], {
+        require!(selector => bool);
+    });
```

Comment:
The lookup merging is great, but I am not sure if I follow why the constraint for `selector` being a boolean can be removed?

Seems like otherwise some attacks could occur, like if we have for example the columns `c1`, `c2`, `e1`, `e2` and want to constrain `e1` being in some range when `c1` and `e2` being in some range when `c2`. Couldn't then the expression `c1 * e1 + c2 * e2` easily be manipulated to land in the prescribed range (while `e1` and `e2` not being in this range at all) by setting `c1` and `c2` to `1` and setting `e1`, `e2` to slightly over `(p-1)/2`?

In fact, I have an impression even an additional constraint would be needed - for each of the conditions being a boolean?

Reply:
Yes you're totally correct. The only way it would be optional is those things are checked elsewhere (either explicitly look the boolean check or implicitly by making sure the lookups are mutually exclusive because e.g. only used once per state in the state machine). For a generalized system these checks would indeed be needed so I guess I should add them, thanks!